[Doc] Add sections about how to enable and configure chunking for producers/consumers (C++ client) (#14063)
apache/pulsar: site2/docs/client-libraries-cpp.md

producerConf.setLazyStartPartitionedProducers(true);
```

### Enable chunking

Message [chunking](concepts-messaging.md#chunking) enables Pulsar to process large payload messages by splitting the message into chunks at the producer side and aggregating chunked messages at the consumer side. 

> The message chunking feature is OFF by default. The following is an example about how to enable message chunking when creating a producer.

Reviewer: ```suggestion
The message chunking feature is OFF by default. The following is an example of how to enable message chunking when creating a producer.
```

### Blocking example

The benefit of this approach is that it is the simplest code. Simply keeps calling `receive(msg)` which blocks until a message is received.

This example starts a subscription at the earliest offset and consumes 100 messages.

```c++
#include <pulsar/Client.h>

using namespace pulsar;

int main() {
    Client client("pulsar://localhost:6650");

    Consumer consumer;
    ConsumerConfiguration config;
    config.setSubscriptionInitialPosition(InitialPositionEarliest);
    Result result = client.subscribe("persistent://public/default/my-topic", "consumer-1", config, consumer);
    if (result != ResultOk) {
        std::cout << "Failed to subscribe: " << result << std::endl;
        return -1;
    }

    Message msg;
    int ctr = 0;
    // consume 100 messages
    while (ctr < 100) {
        consumer.receive(msg);
        std::cout << "Received: " << msg
            << "  with payload '" << msg.getDataAsString() << "'" << std::endl;

        consumer.acknowledge(msg);
        ctr++;
    }

    std::cout << "Finished consuming synchronously!" << std::endl;

    client.close();
    return 0;
}
```

### Consumer with a message listener

> You can avoid  running a loop with blocking calls with an event based style by using a message listener which is invoked for each message that is received.

Reviewer 2: ```suggestion
You can avoid running a loop with blocking calls with an event based style by using a message listener which is invoked for each message that is received.
```

#include <thread>

using namespace pulsar;

std::atomic<uint32_t> messagesReceived;

void handleAckComplete(Result res) {
    std::cout << "Ack res: " << res << std::endl;
}

void listener(Consumer consumer, const Message& msg) {
    std::cout << "Got message " << msg << " with content '" << msg.getDataAsString() << "'" << std::endl;
    messagesReceived++;
    consumer.acknowledgeAsync(msg.getMessageId(), handleAckComplete);
}

int main() {
    Client client("pulsar://localhost:6650");

    Consumer consumer;
    ConsumerConfiguration config;
    config.setMessageListener(listener);
    config.setSubscriptionInitialPosition(InitialPositionEarliest);
    Result result = client.subscribe("persistent://public/default/my-topic", "consumer-1", config, consumer);
    if (result != ResultOk) {
        std::cout << "Failed to subscribe: " << result << std::endl;
        return -1;
    }

    // wait for 100 messages to be consumed
    while (messagesReceived < 100) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    std::cout << "Finished consuming asynchronously!" << std::endl;

    client.close();
    return 0;
}
```

### Configure chunking

> You can limit the maximum number of chunked messages a consumer maintains concurrently by configuring the `setMaxPendingChunkedMessage` and `setAutoOldestChunkedMessageOnQueueFull` parameter. When the threshold is reached, the consumer drops pending messages by silently acknowledging them or asking the broker to redeliver them later. 

Reviewer 3: ```suggestion
You can limit the maximum number of chunked messages a consumer maintains concurrently by configuring the `setMaxPendingChunkedMessage` and `setAutoOldestChunkedMessageOnQueueFull` parameters. When the threshold is reached, the consumer drops pending messages by silently acknowledging them or asking the broker to redeliver them later. 
```

Author: Sharp eyes! Thanks for pointing this out. I've corrected it. PTAL.

> You can limit the maximum number of chunked messages a consumer maintains concurrently by configuring the `setMaxPendingChunkedMessage` and `setAutoOldestChunkedMessageOnQueueFull` parameters. When the threshold is reached, the consumer drops pending messages by silently acknowledging them or asking the broker to redeliver them later. 

Reviewer 2: ```suggestion
You can limit the maximum number of chunked messages a consumer maintains concurrently by configuring the `maxPendingChunkedMessage` and `autoOldestChunkedMessageOnQueueFull` parameters. When the threshold is reached, the consumer drops pending messages by silently acknowledging them or asking the broker to redeliver them later. 
```

I think it would be more reasonable to use `maxPendingChunkedMessage` and `autoOldestChunkedMessageOnQueueFull` here. `setMaxPendingChunkedMessage` and `setAutoOldestChunkedMessageOnQueueFull` are the setting methods for these parameters

Reviewer 2: @BewareMyPower I think there is a typo in `autoOldestChunkedMessageOnQueueFull` parameter of the c++ client. It should be `autoAckOldestChunkedMessageOnQueueFull`. PTAL.

Reviewer 3: Yes. There is a missed `Ack` word. I'll open another PR for it.

> You can limit the maximum number of chunked messages a consumer maintains concurrently by configuring the `setMaxPendingChunkedMessage` and `setAutoOldestChunkedMessageOnQueueFull` parameters. When the threshold is reached, the consumer drops pending messages by silently acknowledging them or asking the broker to redeliver them later. 

Reviewer 3: ```suggestion
You can limit the maximum number of chunked messages a consumer maintains concurrently by configuring the `setMaxPendingChunkedMessage` and `setAutoAckOldestChunkedMessageOnQueueFull` parameters. When the threshold is reached, the consumer drops pending messages by silently acknowledging them or asking the broker to redeliver them later. 
```

This patch should be applied after #14070 is merged.

The following is an example of how to configure message chunking.

```c++

> ConsumerConfiguration conf;
> conf.setAutoOldestChunkedMessageOnQueueFull(true);
> conf.setMaxPendingChunkedMessage(100);
> Consumer consumer;
> client.subscribe("my-topic", "my-sub", conf, consumer);

Reviewer 3: ```suggestion
ConsumerConfiguration conf;
conf.setAutoAckOldestChunkedMessageOnQueueFull(true);
conf.setMaxPendingChunkedMessage(100);
Consumer consumer;
client.subscribe("my-topic", "my-sub", conf, consumer);
```

This patch should be applied after #14070 is merged.
